homeassistant/components/mqtt/__init__.py

```diff
@@ -953,7 +956,9 @@ def _mqtt_handle_message(self, msg) -> None:
         timestamp = dt_util.utcnow()
 
         for subscription in self.subscriptions:
-            if not _match_topic(subscription.topic, msg.topic):
+            try:
+                next(subscription.matcher.iter_match(msg.topic))
+            except StopIteration:
```

Comment:
Also possible:

```suggestion
            if not next(subscription.matcher.iter_match(msg.topic), False):
```

Reply:
but maybe "cleaner" if `_matcher_for_topic` returns

```python
return lambda topic: next(subscription.matcher.iter_match(msg.topic), False)
```

So we get
```suggestion
            if not subscription.matcher(msg.topic):
```